### PR TITLE
tests(smoke): check runner result is sensible before using

### PR DESCRIPTION
**Pull request by connorjclark**
Once #14320 landed I then saw another error further down the report assertion function. So let's fix the root cause: we shouldn't just _trust_ that `lighthouseRunner` returns a valid LH.Result. At least for the CLI runner, it's possible an error occurred (ex: config resolution failed) resulting in no LHR at all (just an empty object being returned. or maybe it was undefined).

### CODE REVIEW

**Comment by brendankenny** [comment]
I believe the intention was `lighthouseRunner` would either return a `Promise<{lhr: LHResult, artifacts: Artifacts, log: string}>` or throw an error. In the error case the `catch` will catch it and continue to the end of the loop, never running the assertions (stderr should be logged and the whole test counted as a failure).

Is the problem that the non-CLI versions are returning _something_ in the error cases instead? The CLI has this check basically built in because it loads the LHR and artifacts from disk, so will throw if they're not around:

https://github.com/GoogleChrome/lighthouse/blob/bb2bed6a5576261fc23fbab0c603988bdbcc15f4/cli/test/smokehouse/lighthouse-runners/cli.js#L101-L110

Can we do the same in bundle/devtools?

We could loosen the types for `lighthouseRunner` instead, but `bundle` already does [a third of the check here](https://github.com/GoogleChrome/lighthouse/blob/bb2bed6a5576261fc23fbab0c603988bdbcc15f4/cli/test/smokehouse/lighthouse-runners/cli.js#L101-L110), and `devtools` is [writing possibly nonsense to disk here](https://github.com/GoogleChrome/lighthouse/blob/bb2bed6a5576261fc23fbab0c603988bdbcc15f4/cli/test/smokehouse/lighthouse-runners/devtools.js#L60-L61), so it seems like a benefit. In theory the check can grow to runner-specific checks like cli does as well (like [checking](https://github.com/GoogleChrome/lighthouse/blob/bb2bed6a5576261fc23fbab0c603988bdbcc15f4/cli/test/smokehouse/lighthouse-runners/cli.js#L118-L123) that if there's a non-zero exit code there was also a `lhr.runtimeError` in the expectations), so e.g. `devtools` could log more specific errors due to build issues or pptr setup or whatever.

File: cli/test/smokehouse/smokehouse.js [hunk 1]
```diff
@@ -195,6 +195,13 @@ async function runSmokeTest(smokeTestDefn, testOptions) {
         ...await lighthouseRunner(requestedUrl, configJson, {isDebug, useLegacyNavigation}),
         networkRequests: takeNetworkRequestUrls ? takeNetworkRequestUrls() : undefined,
       };
+
+      if (!result || !result.lhr?.audits || !result.artifacts) {
+        // Don't try again–something went really wrong! ex: the config could not be resolved
+        i = retries + 1;
+        throw new Error('lighthouse runner returned a bad result. got lhr:\n' +
+          JSON.stringify(result.lhr, null, 2));
+      }
```
Comment (paulirish) [inline cli/test/smokehouse/smokehouse.js:204]:
```suggestion
      if (result.lhr?.runtimeError?.code === 'UNKNOWN_ERROR') {
        // Don't try again–something went really wrong! ex: the config could not be resolved
        i = retries + 1;
        throw new Error('lighthouse runner returned a bad result. got lhr:\n' +
          JSON.stringify(result.lhr, null, 2));
      }
```

Comment (brendankenny) [inline cli/test/smokehouse/smokehouse.js:204]: "just remove all the bugs" is a fair point :)

What about aiming for runner-specific checking in the runner, but keep a generic check here? e.g.

```suggestion
      if (!result.lhr?.audits || !result.artifacts) {
        // Something went really wrong and the runner didn't catch it.
        throw new Error('lighthouse runner returned a bad result. got lhr:\n' +
          JSON.stringify(result.lhr, null, 2));
      }
```

Skipping the retries does seem unnecessary (especially by manipulating `i` :P); if it's that bad and it's happening outside of debugging we really do need to fix the runner

**Comment by paulirish** [comment]
mkay

**Review by paulirish — APPROVED** [review]

**Review by brendankenny — APPROVED** [review]

**Pull request merged**